tests/returntypes/gradients/test_pulse_gradient_new.py

```diff
@@ -0,0 +1,900 @@
+        params_0 = [jnp.array(1.51), jnp.array(-0.371)]
+        params_1 = [jnp.array([0.2, 0.2, -0.4])]
+        dev = qml.device("default.qubit.jax", wires=2)
+
+        @qml.qnode(dev, interface="jax")
+        def qnode(params_0, params_1):
+            qml.evolve(ham_0)(params_0, 0.2)
+            qml.evolve(ham_1)(params_1, 0.4)
+            return qml.expval(qml.PauliY(0) @ qml.PauliZ(1))
+
+        qnode.construct((params_0, params_1), {})
+
+        num_samples = 50
+        qnode.tape.trainable_params = [0, 1, 2]
+        tapes, fn = stoch_pulse_grad(qnode.tape, num_samples=num_samples, sampler_seed=7123)
+        assert len(tapes) == 3 * 2 * num_samples
+
+        res = fn(qml.execute(tapes, dev, None))
+        exp_grad = jax.grad(qnode, argnums=(0, 1))(params_0, params_1)
+        exp_grad = exp_grad[0] + exp_grad[1]
+        assert all(qml.math.allclose(r, e, rtol=0.2) for r, e in zip(res, exp_grad))
+
+    def test_with_jit(self):
+        """Test that the stochastic parameter-shift rule works with JITting."""
+        import jax
+
+        jax.config.update("jax_enable_x64", True)
+        dev = qml.device("default.qubit.jax", wires=1)
+        jnp = jax.numpy
+        T = (0.2, 0.5)
+        ham_single_q_const = qml.pulse.constant * qml.PauliY(0)
+
+        def fun(params):
+            op = qml.evolve(ham_single_q_const)(params, T)
+            tape = qml.tape.QuantumScript([op], [qml.expval(qml.PauliZ(0))])
+            tapes, fn = stoch_pulse_grad(tape)
+            assert len(tapes) == 2
+            res = fn(qml.execute(tapes, dev, None))
+            return res
+
+        params = [jnp.array(0.24)]
+        # Effective rotation parameter
+        p = params[0] * (T[1] - T[0])
+        res = fun(params)
+        assert qml.math.isclose(res, -2 * jnp.sin(2 * p) * (T[1] - T[0]))
+        res_jit = jax.jit(fun)(params)
+        assert qml.math.isclose(res, res_jit)
+
+
+@pytest.mark.jax
+class TestStochPulseGradQNodeIntegration:
+    """Test that stoch_pulse_grad integrates correctly with QNodes."""
+
+    @pytest.mark.parametrize("shots, tol", [(None, 1e-4), (100, 0.1), ([100, 99], 0.1)])
+    @pytest.mark.parametrize("num_samples", [1, 3])
+    def test_simple_qnode_expval(self, num_samples, shots, tol):
+        """Test that a simple qnode that returns an expectation value
+        can be differentiated with stoch_pulse_grad."""
+        import jax
+
+        jnp = jax.numpy
+        jax.config.update("jax_enable_x64", True)
+        dev = qml.device("default.qubit.jax", wires=1, shots=shots, prng_key=jax.random.PRNGKey(74))
+        T = 0.2
+        ham_single_q_const = qml.pulse.constant * qml.PauliY(0)
+
+        @qml.qnode(dev, interface="jax", diff_method=stoch_pulse_grad, num_samples=num_samples)
+        def circuit(params):
+            qml.evolve(ham_single_q_const)(params, T)
+            return qml.expval(qml.PauliZ(0))
+
+        params = [jnp.array(0.4)]
+        grad = jax.jacobian(circuit)(params)
```

Comment:
why jacobian?

Reply:
That... is a good question. Will make it `grad`.

Reply:
Ah nevermind: The shot vectors make `jacobian` necessary. My bad :facepalm:

Reply:
Ah right the shot vector! ✔️